### DISCUSSION TITLE
gNB error: exceeded maximum number of timed out transmissions

### DISCUSSION THREAD

**Discussion by initChat** [srsran/srsRAN_Project#550] [General Help] [answered]
I have USRP B210. I run the open5gs docker as the core network. The gNB can connect but it always returns `Error: exceeded maximum number of timed out transmissions.` what did I do wrong? Thank you very much.

Below is my config files `gnb_rf_b210_fdd_srsUE.yml`. And how I run code (open5gs docker -> performance script -> gNB with config files). And i attacted my gnb log below

gnb_rf_b210_fdd_srsUE.yml
```
amf:
  addr: 10.53.1.2                                                 # The address or hostname of the AMF.
  bind_addr: 10.53.1.1                                            # A local IP that the gNB binds to for>

ru_sdr:
  device_driver: uhd                                              # The RF driver name.
  device_args: type=b200                                          # Optionally pass arguments to the selected RF driver.
  clock: internal
  sync:                                                           # (@!mm - changed from `external` to `internal`) Set sync to external. This set-up uses a LEO BODNAR GPDSO providing a 10 MHz ref.
  srate: 23.04                                                    # RF sample rate might need to be adjusted according to selected bandwidth.
  tx_gain: 89                                                     # Transmit gain of the RF might need to adjusted to the given situation.
  rx_gain: 76                                                     # Receive gain of the RF might need to adjusted to the given situation.

cell_cfg:
  dl_arfcn: 368500                                                # ARFCN of the downlink carrier (center frequency).
  band: 3                                                         # The NR band.
  channel_bandwidth_MHz: 20                                       # Bandwith in MHz. Number of PRBs will be automatically derived.
  common_scs: 15                                                  # Subcarrier spacing in kHz used for data.
  plmn: "00101"                                                   # PLMN broadcasted by the gNB.
  tac: 7                                                          # Tracking area code (needs to match the core configuration). 
  pdcch:
    dedicated:
      ss2_type: common                                            # Set search space to common to match capabilities
      dci_format_0_1_and_1_1: false                               # Use fallback DCI to match srsUE capabilities
    common:
      ss0_index: 0                                                # Set search space zero index to match srsUE capabilities
      coreset0_index: 12                                          # Set search CORESET Zero index to match srsUE capabilities
  prach:
    prach_config_index: 1                                         # Set PRACH config index to match srsUE expectation 

log:
  filename: /tmp/gnb.log                                          # Path of the log file.
  all_level: info                                                 # Logging level applied to all layers.
```
I run the open5gs docker
```
# run open5gs
>> cd srsRAN_Project/docker
>> sudo docker compose up --build 5gc
```
then run the performance script
```
# run performance script
>> echo "performance" | sudo tee /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
>> cd srsRAN_Project/scripts
>> ./srsran_performance 
```
after that i run the gnb code with config files
```
sudo ./gnb -c gnb_rf_b210_fdd_srsUE.yml 

The PRACH detector will not meet the performance requirements with the configuration {Format 0, ZCZ 0, SCS 1.25kHz, Rx ports 1}.
Lower PHY in quad executor mode.

--== srsRAN gNB (commit 2f90c8b60) ==--

Connecting to AMF on 10.53.1.2:38412
Available radio types: uhd and zmq.
[INFO] [UHD] linux; GNU C++ version 10.5.0; Boost_107400; UHD_4.0.0.HEAD-0-g90ce6062
[INFO] [LOGGING] Fastpath logging disabled at runtime.
Making USRP object with args 'type=b200'
[INFO] [B200] Detected Device: B210
[INFO] [B200] Operating over USB 3.
[INFO] [B200] Initialize CODEC control...
[INFO] [B200] Initialize Radio control...
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Setting master clock rate selection to 'automatic'.
[INFO] [B200] Asking for clock rate 16.000000 MHz... 
[INFO] [B200] Actually got clock rate 16.000000 MHz.
[INFO] [MULTI_USRP] Setting master clock rate selection to 'manual'.
[INFO] [B200] Asking for clock rate 23.040000 MHz... 
[INFO] [B200] Actually got clock rate 23.040000 MHz.
Cell pci=1, bw=20 MHz, dl_arfcn=368500 (n3), dl_freq=1842.5 MHz, dl_ssb_arfcn=368410, ul_freq=1747.5 MHz

==== gNodeB started ===
Type <t> to view trace
Late: 0; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 2796; Underflow: 1; Overflow: 1;
Error: exceeded maximum number of timed out transmissions.
Late: 6000; Underflow: 0; Overflow: 0;
Error: exceeded maximum number of timed out transmissions.
Late: 5988; Underflow: 2; Overflow: 0;
Error: exceeded maximum number of timed out transmissions.
Late: 6000; Underflow: 0; Overflow: 0;
Error: exceeded maximum number of timed out transmissions.
```
my gnb.log 
[gnb.log](https://github.com/srsran/srsRAN_Project/files/14929192/gnb.log)

**Comment by pgawlowicz** [comment]
is your PC powerful enough? Could you provide CPU and memory specs?

**Reply by initChat** [reply]
This is my CPU and memory. If I present wrong info please let me know. Thank you very much. 
```
CPU(s):                  8
model name	: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz


>> free
                  total               used          free             shared       buff/cache   available
Mem:        16016184     4679964     6372876     1009740     4963344     9987012
Swap:       31999996           0            31999996

>> df -h
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           1.6G  2.9M  1.6G   1% /run
/dev/nvme0n1p8  166G   50G  108G  32% /
tmpfs           7.7G  464K  7.7G   1% /dev/shm
tmpfs           5.0M  4.0K  5.0M   1% /run/lock
efivarfs        246K  149K   93K  62% /sys/firmware/efi/efivars
/dev/nvme0n1p6  921M  197M  661M  23% /boot
/dev/nvme0n1p1   96M   85M   12M  89% /boot/efi
tmpfs           1.6G  3.2M  1.6G   1% /run/user/1000
```

**Reply by pgawlowicz** [reply]
ok, the PC is good enough. Could you power cycle b210 and load the image again? Did you try using different version of uhd?

**Reply by initChat** [reply]
I try change the version into 3.15 as mentioned in [rf driver](https://docs.srsran.com/projects/project/en/latest/user_manuals/source/installation.html) but when I `git checkout v.3.15.0.0` it shows error. 
So I checkout to `v4.1.0.0`. And then do the install again, and download the image again, the error still presented.
What should I do?

**Reply by pgawlowicz** [reply]
could you try to set the sync field in the config to internal? do you have an external clock to try?

**Reply by initChat** [reply]
I don't have external clock to try.
I change the sync to internal, and it still have the same problem
```
ru_sdr:
  device_driver: uhd                                            
  device_args: type=b200                                          
  clock: internal
  sync: internal   
```

**Reply by pgawlowicz** [reply]
could you try to test the usrp with the UHD tools:
/usr/local/lib/uhd/examples/benchmark_rate --args type=b200 --rx_rate 23.04e6 --tx_rate 23.04e6
/usr/local/lib/uhd/examples/latency_test
/usr/local/lib/uhd/examples/test_timed_commands
/usr/local/lib/uhd/examples/tx_waveforms --freq=900e6 --rate=23.04e6

**Reply by initChat** [reply]
all example tests is worked, I have attached results below. 
I decided to `make` the sraRAN_Project again as I see the UHD version in gnb screen still said `v4.0`. After `make`, the version is said v4.1.0.
```
--== srsRAN gNB (commit 2f90c8b60) ==--

Connecting to AMF on 10.53.1.2:38412
Available radio types: uhd and zmq.
[INFO] [UHD] linux; GNU C++ version 10.5.0; Boost_107400; UHD_4.1.0.HEAD-0-gd21735d5
```
However, the problem still happended.
[gnb.log](https://github.com/srsran/srsRAN_Project/files/14931188/gnb.log)



[benchmark.txt](https://github.com/srsran/srsRAN_Project/files/14931155/benchmark.txt)
[latency_test.txt](https://github.com/srsran/srsRAN_Project/files/14931156/latency_test.txt)
[timeout.txt](https://github.com/srsran/srsRAN_Project/files/14931157/timeout.txt)
[tx_waveform.txt](https://github.com/srsran/srsRAN_Project/files/14931158/tx_waveform.txt)

**Reply by pgawlowicz** [reply]
hmm, ok, what is your system? could you check CPU load when runing gnb? with htop

**Comment by initChat** [comment]
It will show some core has 100%. Does this means my computer is not sufficient to run the code? Is there any solution? Thank you

![Screenshot from 2024-04-10 18-11-30](https://github.com/srsran/srsRAN_Project/assets/40858099/c5a0a55d-46e9-47dd-9496-079b12a8c65e)
![Screenshot from 2024-04-10 18-11-35](https://github.com/srsran/srsRAN_Project/assets/40858099/32b43c37-0881-4faa-95c9-d97465c870df)
![Screenshot from 2024-04-10 18-12-21](https://github.com/srsran/srsRAN_Project/assets/40858099/2799f3fc-1d14-42fe-9452-e5c6c697297f)

**Reply by pgawlowicz** [reply]
hmm, not sure why gnb operates only with a single CPU core. @frankist any hints?

**Reply by initChat** [reply]
I checked that if the gnb operate on only one core, however it operate for core 0 - 7
```
>> sudo netstat -tupln | grep gnb
udp        0      0 10.53.1.1:2152          0.0.0.0:*                           49425/./gnb      

>> taskset -pc 0-7 49425
pid 49425's current affinity list: 0-7
```

What should I do?

**Reply by pgawlowicz** [reply]
What OS do you use? do you run gnb in docker? Did you maybe play with `expert_execution` in the gnb config?

**Reply by initChat** [reply]
I run on Ubuntu 22.04, and not gnb in docker. i install srsRAN_Project from source, and no expert_execution in gnb config

**Reply by pgawlowicz** [reply]
could you start with `sudo taskset -c 0-7 ./gnb -c ./your_config_file`

**Reply by initChat** [reply, marked as answer]
I decide to **erase and reinstall ubuntu 22.04 again**. And install everything again, and it can works now! I can run the gNB for 10 mins right now, will test on longer time later.
Sometimes it still shows snapshot of `htop` as 100% in some one core. But the trace of gNB is working well, shows no problem.
Thank you very much

```
==== gNodeB started ===
Type <t> to view trace
Late: 19; Underflow: 9; Overflow: 0;
Late: 0; Underflow: 2; Overflow: 0;
Late: 10; Underflow: 3; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 8; Underflow: 5; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 10; Underflow: 2; Overflow: 0;
Late: 5; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 5; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 5; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 2; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
Late: 0; Underflow: 4; Overflow: 0;
Late: 0; Underflow: 1; Overflow: 0;
```
![Screenshot 2024-04-12 at 17-19-17 stdin](https://github.com/srsran/srsRAN_Project/assets/40858099/48ec5eef-a783-4f07-b2d5-7182e4bbf0de)
![Screenshot 2024-04-12 at 17-18-17 stdin](https://github.com/srsran/srsRAN_Project/assets/40858099/d155ec3e-74a7-47ca-a717-f71316edf89d)
![Screenshot 2024-04-12 at 17-16-44 stdin](https://github.com/srsran/srsRAN_Project/assets/40858099/21ae2419-fd6e-4fe4-9e5a-463650445034)

**Reply by pgawlowicz** [reply]
Thanks for the feedback!

**Comment by houshangaziziaccelleran** [comment]
Hi
Is this issue already fixed? I have the same issue with my srsran/n310
